woocommerce/google-listings-and-ads
# Update Google API v16

This PR: 

- updates the Google API to v16 
- upgrades all its dependencies
- Tweak the code based on the API changes
- Tweak the tests 
- Tweak CleanUpServices (structural changes required so)
- Fixes some bugs and deleted unused code

Please refer to the file comments to get context of all the changes. But in brief:

- Run `composer update googleads/google-ads-php --with-all-dependencies`
- Update V14 references to V16
- Replace sales_country with feed_label
- Replace Vertical with Source
- Add Request as param in some Mutation calls instead of ID+Operation
- Replace MerchantLink logic (they replaced it with ProductLink)
- Add /Client/ subfolder in the CleanupService
- Add an option to exclude deletions of certain services

```
composer update googleads/google-ads-php --with-all-dependencies

Upgrading googleads/google-ads-php (v19.2.0 => dev-legacy-v22.1.0 905fd78)
  - Upgrading google/auth (v1.34.0 => v1.37.1)
  - Upgrading google/common-protos (v4.4.0 => v4.5.0)
  - Upgrading google/gax (v1.25.0 => v1.29.1)
  - Upgrading google/longrunning (v0.2.6 => 0.4.1)
  - Upgrading google/protobuf (v3.25.0 => v3.25.3)
  - Upgrading monolog/monolog (2.9.2 => 2.9.3)
  - Upgrading symfony/deprecation-contracts (v2.5.2 => v2.5.3)
  - Upgrading symfony/polyfill-ctype (v1.28.0 => v1.29.0)
  - Upgrading symfony/polyfill-intl-normalizer (v1.28.0 => v1.29.0)
  - Upgrading symfony/polyfill-mbstring (v1.28.0 => v1.29.0)
  - Upgrading symfony/polyfill-php80 (v1.28.0 => v1.29.0)
  - Upgrading symfony/polyfill-php81 (v1.28.0 => v1.29.0)
```

```
composer update google/apiclient--with-all-dependencies
Upgrading google/apiclient (v2.15.3 => v2.16.0)
  - Upgrading google/apiclient-services (v0.331.0 => v0.350.0)
```

### Steps for testing

1. Checkout this PR
2. Run `rm -rf vendor && composer install`
3. Perform smoke testing as defined here https://github.com/woocommerce/google-listings-and-ads/wiki/Smoke-tests
4. Check there are no references to V14 in the code
5. Double check with https://developers.google.com/google-ads/api/docs/upgrade (v14 to v15 and v15 to v16) to see if we cover all the cases affecting us.
6. See[ E2E tests passing](https://github.com/woocommerce/google-listings-and-ads/actions/runs/8873455703) and [Unit Tests Passing](https://github.com/woocommerce/google-listings-and-ads/actions/runs/8873407567)

### Additional details

After merging this PR we should bump a minor version instead of a patch. So 2.7.1

Related to https://github.com/Automattic/woocommerce-connect-server/pull/2460

### Changelog entry

> Update - Update Google API to V16
